From fcdddf2b726439e049992878f90da607414a1a47 Mon Sep 17 00:00:00 2001
From: Denis Kenzior <denkenz@gmail.com>
Date: Mon, 3 Feb 2020 11:54:28 -0600
Subject: [PATCH] build: Support missing rawmemchr

rawmemchr is a GNU extension in glibc that does not exist in musl.

Upstream-status: Backport of https://git.kernel.org/pub/scm/network/wireless/iwd.git/commit/?id=fcdddf2b726439e049992878f90da607414a1a47

Signed-off-by: Robert Joslyn <robert.joslyn@redrectangle.org>

---
 configure.ac  |  1 +
 src/missing.h | 10 ++++++++++
 src/wiphy.c   |  1 +
 3 files changed, 12 insertions(+)

diff --git a/configure.ac b/configure.ac
index 5ae1401cae17..2d373a47ba68 100644
--- a/configure.ac
+++ b/configure.ac
@@ -128,6 +128,7 @@ AC_DEFINE_UNQUOTED(WIRED_STORAGEDIR, "${wired_storagedir}",
 			[Directory for Ethernet daemon storage files])
 
 AC_CHECK_FUNCS(explicit_bzero)
+AC_CHECK_FUNCS(rawmemchr)
 
 AC_CHECK_HEADERS(linux/types.h linux/if_alg.h)
 
diff --git a/src/missing.h b/src/missing.h
index 2bb210ae3c81..2cc80aee5d38 100644
--- a/src/missing.h
+++ b/src/missing.h
@@ -27,3 +27,13 @@ static inline void explicit_bzero(void *s, size_t n)
         __asm__ __volatile__ ("" : : "r"(s) : "memory");
 }
 #endif
+
+#ifndef HAVE_RAWMEMCHR
+static inline void *rawmemchr(const void *s, int c)
+{
+_Pragma("GCC diagnostic push")
+_Pragma("GCC diagnostic ignored \"-Wstringop-overflow=\"")
+	return memchr(s, c, (size_t) -1);
+_Pragma("GCC diagnostic pop")
+}
+#endif
diff --git a/src/wiphy.c b/src/wiphy.c
index 1da479db2dab..511bb27f52b8 100644
--- a/src/wiphy.c
+++ b/src/wiphy.c
@@ -37,6 +37,7 @@
 
 #include "linux/nl80211.h"
 
+#include "src/missing.h"
 #include "src/iwd.h"
 #include "src/module.h"
 #include "src/ie.h"
-- 
2.21.0

